File: ABC/401-450/404/d.py
import sys
sys.setrecursionlimit(10**6)

def main():
  n,m = map(int,input().split())
  c = list(map(int,input().split()))
  k = []
  a = [] 
  zoo_to_animal = [[] for _ in range(n)]
  for i in range(m):
    ka = list(map(int,input().split()))
    ki = ka[0]
    ai = ka[1:]
    k.append(ki)
    ai = [x-1 for x in ai]
    a.append(ai)
    for x in ai:
      zoo_to_animal[x].append(i)
  ans = [1001001001]
  def f(d,now,cnt):
    if d == n:
      for i in range(m):
        if cnt[i] < 2:
          return
      ans[0] = min(ans[0],now)
      return
    for i in range(3):
      cnt2 = cnt.copy()
      now2 = now + c[d] * i
      for x in zoo_to_animal[d]:
        cnt2[x] += i
      f(d+1,now2,cnt2)
      
  f(0,0,[0]*m)
  print(ans[0])


if __name__ == '__main__':
  main()

File: ABC/401-450/404/d.cpp
#include <iostream>
#include<vector>
#include<algorithm>
using namespace std;

int main(){
  int n,m;
  cin >> n >> m;
  vector<long long> c(n);
  for(int i=0;i<n;i++) cin >> c[i];
  vector<int> k(m);
  vector<vector<int> > a(m);
  vector<vector<int> > b(n);
  for(int i=0;i<m;i++){
    cin >> k[i];
    for(int j=0;j<k[i];j++){
      int ai;
      cin >> ai;
      ai--;
      a[i].push_back(ai);
      b[ai].push_back(i);
    }
  }
  long long ans = 1e18;
  auto dfs = [&](auto dfs, int d, vector<int> cnt,long long now)->void{
    if(d == n){
      for(int i=0;i<m;i++){
        if(cnt[i] < 2) return;
      }
      ans = min(ans,now);
      return;
    }
    for(int i=0;i<3;i++){
      vector<int> cnt2=cnt;
      for(int j=0;j<b[d].size();j++){
        cnt2[b[d][j]] += i;
      }
      long long now2 = now + (long long)i*c[d];
      dfs(dfs,d+1,cnt2,now2);
    }
  };
  dfs(dfs,0,vector<int>(m),0LL);
  cout << ans << endl;
  return 0;
}

